dali/python/nvidia/dali/plugin/pytorch.py

```diff
@@ -241,57 +230,41 @@ def __next__(self):
-                output[-1][category] = output[-1][category][0:data_fromlastGPU]
-            return output
+        if self._reader_name:
+            self._counter += self.batch_size
+            if_drop, left = self._remove_padded()
+            if np.any(if_drop):
+                output = []
+                for batch, to_copy in zip(self._data_batches, left):
+                    batch = batch.copy()
+                    for category in self._output_categories:
+                        batch[category] = batch[category][0:to_copy]
+                    output.append(batch)
+                return output
 
-        return self._data_batches
-
-    def next(self):
-        """
-        Returns the next batch of data.
-        """
-        return self.__next__()
-
-    def __iter__(self):
-        return self
-
-    def reset(self):
-        """
-        Resets the iterator after the full epoch.
-        DALI iterators do not support resetting before the end of the epoch
-        and will ignore such request.
-        """
-        if self._counter >= self._size or self._size < 0:
-            if self._fill_last_batch and not self._last_batch_padded:
-                self._counter = self._counter % self._size
-            else:
-                self._counter = 0
-            for p in self._pipes:
-                p.reset()
-                if p.empty():
-                    with p._check_api_type_scope(types.PipelineAPIType.ITERATOR):
-                        p.schedule_run()
         else:
-            logging.warning("DALI iterator does not support resetting while epoch is not finished. Ignoring...")
+            self._counter += self._num_gpus * self.batch_size
+            if (not self._fill_last_batch) and (self._counter > self._size) and self._size > 0:
+                # First calculate how much data is required to return exactly self._size entries.
+                diff = self._num_gpus * self.batch_size - (self._counter - self._size)
+                # Figure out how many GPUs to grab from.
+                numGPUs_tograb = int(np.ceil(diff/self.batch_size))
+                # Figure out how many results to grab from the last GPU (as a fractional GPU batch may be required to
+                # bring us right up to self._size).
+                mod_diff = diff % self.batch_size
+                data_fromlastGPU = mod_diff if mod_diff else self.batch_size
+
+                # Grab the relevant data.
+                # 1) Grab everything from the relevant GPUs.
+                # 2) Grab the right data from the last GPU.
+                # 3) Append data together correctly and return.
+                output = self._data_batches[0:numGPUs_tograb]
+                output[-1] = output[-1].copy()
+                for category in self._output_categories:
+                    output[-1][category] = output[-1][category][0:data_fromlastGPU]
+                return output
+
+        return self._data_batches
```

Comment:
I feel like I see the same thing for the third time. Can we move this or most of it to the base class? I know there is this slight difference in iterating over some range or list of categories, but still.

Reply:
The problem with this code is that it is not very accurate (compared to the new implementation), and it only PyTorch and PaddlePaddle implementations match, while MXNet is a bit different.
I don't know if I should touch it as this code path is not tested anyway.

Reply:
Hmm, Maybe it will be worth to think about deprecation warnings than. If this is the old, untested way of doing things, and we do not change it, let's leave it.

Reply:
Added